frame/contracts/src/wasm/code_cache.rs

```diff
@@ -185,14 +199,50 @@ fn increment_64(refcount: &mut u64) {
 	");
 }
 
-/// Token to be supplied to the gas meter which charges the weight needed for reinstrumenting
-/// a contract of the specified size in bytes.
+/// Get the size of the instrumented code stored at `code_hash` without loading it.
+///
+/// The returned value is slightly too large because it also contains the fields apart from
+/// `code` which are located inside [`PrefabWasmModule`]. However, those are negligible when
+/// compared to the code size. Additionally, charging too much weight is completely safe.
+fn estimate_code_size<T: Config>(code_hash: &CodeHash<T>) -> Result<u32, DispatchError>
+where
+	T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>
+{
+	let key = <CodeStorage<T>>::hashed_key_for(code_hash);
+	let mut data = [0u8; 0];
+	let len = sp_io::storage::read(&key, &mut data, 0).ok_or_else(|| Error::<T>::CodeNotFound)?;
+	Ok(len)
+}
+
+/// Costs for operations that are related to code handling.
 #[cfg_attr(test, derive(Debug, PartialEq, Eq))]
 #[derive(Clone, Copy)]
-struct InstrumentToken(u32);
+enum CodeToken {
+	/// Weight for instrumenting a contract contract of the supplied size in bytes.
+	Instrument(u32),
+	/// Weight for loading a contract per kilobyte.
+	Load(u32),
+	/// Weight for changing the refcount of a contract per kilobyte.
+	UpdateRefcount(u32),
+}
 
-impl<T: Config> Token<T> for InstrumentToken {
+impl<T> Token<T> for CodeToken
+where
+	T: Config,
+	T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>
+{
 	fn weight(&self) -> Weight {
-		T::WeightInfo::instrument(self.0 / 1024)
+		use self::CodeToken::*;
+		// In case of `Load` and `UpdateRefcount` we already covered the general costs of
+		// accessing the storage but still need to account for the actual size of the
+		// contract code. This is why we substract `T::*::(0)`. We need to do this at this
+		// point because when charging the general weight we do not know the size of
+		// the contract.
+		match *self {
+			Instrument(len) => T::WeightInfo::instrument(len / 1024),
+			Load(len) => T::WeightInfo::code_load(len / 1024) - T::WeightInfo::code_load(0),
+			UpdateRefcount(len) =>
+				T::WeightInfo::code_refcount(len / 1024) - T::WeightInfo::code_refcount(0),
```

Comment:
```suggestion
				T::WeightInfo::code_refcount(len / 1024).saturating_sub(T::WeightInfo::code_refcount(0)),
```
or a comment why it's safe :-)